Update Workflow.ipynb

stscieisenhamer opened the pull request:
Modified the shebang to a more generic form. This form will use whatever python is in the user's path, making it more general. There are some minor drawbacks, but this is a bit more resilent to environment changes.

stscieisenhamer committed:
Update Workflow.ipynb

Modified the shebang to a more generic form. This form will use whatever python is in the user's path, making it more general. There are some minor drawbacks, but this is a bit more resilent to environment changes.

eteq commented:
I went back and forth on this - I agree with you _in general_ @stscieisenhamer, but in this particular form I think @ivastar was using it to illustrate that you can point it at a _specific_ python interpreter.

Do you have any opinion on this, @ivastar ?

stscieisenhamer commented:
The one reason I noted this was due to the use of virtual environments; if they are pushed, which they very well should be, then having the generic version will ensure that scripts always work. Let the environments control what you are doing; the factoid that you can be specific about it in a script would classify as an "advanced topic", ignoring that it really isn't all that advanced.  That is all.

jhunkeler commented:
Maybe classify the generic shebang as a best practice then explain defining an absolute path to the interpreter is non-standard, should be avoided, but is not technically wrong... It just makes **sharing** and **managing** executable scripts a massive pain.

ivastar commented:
I think both are valid and should be illustrated in the notebook. The point I was making is that you can choose a specific python interpreter, partly inspired by troubleshooting with someone two weeks ago - they had a Python 2.7 script that was now incompatible with the Anaconda python and still wanted to be able to run it until they were able to convert it. And yes, I agree with all the points above that you should let the environment variable control this, but in this particular case I say "teach the controversy".

eteq commented:
OK, "teach the controversy" sounds good to me.  @stscieisenhamer, do you think you can update this to reflect these points (both yours and @jhunkeler's?) while still illustrating that this _can_ be done as @ivastar suggests?

sosey commented:
@ivastar @eteq You might consider making sure they now how to create and use python2.7 environment on their machine. It sounds like the script isn't converted for 3.0, so they should be running it in a compatible 2.7 env. It might be better for capatibility for them keep the generic shebang, make sure the 2.7 python they want is used in the environment, and add an extra check into their script to pull the python version - which they can then use to either issue a warning about incompatibility, or keep the script both 2.7 and 3.0 compatible for a while when it's converted .  The check would be something like:

if sys.version_info.major < 3:
    PY3 = False
else:
    PY3 = True

pllim commented:
Alternately, you can also use `six.PY2` or `six.PY3`.

eteq commented:
We've been trying to avoid the py 2.x/3.x waters mostly by just doing py 3.x for all the examples... but that said, some of them do have commented bits like "if you're in py 2.x, do this"...

another option is to go the  `#!/usr/bin/env python3` route?  I _think_ that's standard in most installations, right?

eteq commented:
Oh, and just to be clear, the comment above is motivated partly by the fact that a pretty large share of the audience are people who are pretty new to python, so "information overload" is one of the biggest concerns.

Maybe a sub-slide that details this a bit more?

pllim commented:
If you want to cover IRAF/PyRAF at some point, you will  have to switch back to Python 2. Something to keep in mind when you only give Python 3 examples.

eteq commented:
That's a good point, @pllim... But I'm perfectly happy saying "OK, for this you're stuck with py 2.x... But just put `from __future__ import division, print_function` at the top and that'll pretty much work"...?  I _think_ there's not much in the PyRAF/IRAF workflow that depends on the unicode/str distinctions that are the other main 2/3 issues, right?

One of my main justifications for using py3 is that it's more consistent and understandable to go from 3->2 than the other way around. And scientists being scientists, I think spending any more time than necessary on 2.x examples will make the 3.x switch over a lot harder in the long run...

pllim commented:
PyRAF works in Python 3. But I think somehow IRAF only works in Python 2? Maybe @jhunkeler can say for sure.

sosey commented:
yah, just don't ever cover IRAF, show them how to do what they want in python ;P  I'm also against trying to directly translate equivalencies from IDL/IRAF to Python, in favor of teaching the pythonic way to think about the problem. 

@eteq, yah that may be the standard thing for a python3 shebang. 

Of course you and @ivastar are in the best position to judge your audience, but I'm a fan of teaching them the proper way to do things from the start because people will generally stick with the first thing they learn, go off and write a bunch of scripts like that, and then when they have to start editing them because they're used to virtual envs now, or they are sharing their scripts, it becomes a PITA.  If everyone is learning python from the start, then totally stick with 3.0. But if you are helping people migrate from 2.x to 3.x I think it's worth showing them how to use the environments in the best way currently possible, which will help them avoid PITA code updates as python and the packages they use evolve. 

Also important for them to learn is how to make sure they are maintaining clean environments. It's going to be very easy for them to install  astroconda, and then start adding and changing packages which could mess with dependencies if they aren't careful.  Setting up clean environments for different projects, and knowing how to check what's inside them,  is probably much more important now for users to understand, and will help the people who are using python2 currently to migrate to python3

pllim commented:
PITA... You just spoiled Hunger Games for me!

sosey commented:
@pllim PyRAF is pure python and I believe bernie/chris updated it for py3. IRAF, and more specifically stsdas, has python2.7 specific code and hasn't been updated because we've been moving away from IRAF for what seems like forever now.

sosey commented:
@pllim  no no, that's PEETA :)

pllim commented:
I also wish people would move away from IRAF. But in case you get desperate, instructions to install IRAF with conda is at http://astroconda.readthedocs.io/en/latest/installation.html#installation-with-legacy-iraf-support

stscieisenhamer committed:
Update to match better with the original presentation

Basically separated out the ideas of running a script,
script-as-command, and the modification
of using script-as-command.

ivastar commented:
I'm gonna put my foot down here - we are NOT doing any IRAF during PyLunch.

stscieisenhamer commented:
Getting back to the slide itself, I've pushed a modification that better matches @ivastar intent, while still introducing the more generic form, separating the ideas out just a bit.

eteq commented on session2/Workflow.ipynb:
```diff
@@ -58,28 +58,41 @@
    "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If a script needs to be run by a specific version of python, and not the one normally found when used on the command line, you can specify the full path to that python executable:\n",
```
maybe "used on the command line"-> "when you type `python` on the command line"?

eteq commented on session2/Workflow.ipynb:
````diff
@@ -58,28 +58,41 @@
+    "#!/Users/myusername/anaconda/envs/base-py2/bin/python\n",
+    "```\n",
+    "\n",
+    "To learn more about this Unix magic, lookup [shebang](https://en.wikipedia.org/wiki/Shebang_(Unix)"
````
I think you need a second parenthesis here - the link goes to an error page because it thinks it's `https://en.wikipedia.org/wiki/Shebang_(Unix` rather than `https://en.wikipedia.org/wiki/Shebang_(Unix)`

eteq commented:
These changes look good to me, @stscieisenhamer, aside from a few comments I left above, and one other thing:  the cell you added doesn't seem to have any slideshow metadata.  Did you mean to do that?  The addition you added probably makes sense as a sub-slide (i.e., a thing you get by hitting the down arrow key in the reveal.js show).  To add that just click on the slideshow drop-down in the upper left of the jupyter interface and pick "Sub-slide" (or copy-and-paste the relevant metadata from elsewhere if you just want to edit the json directly).

@ivastar, look good to you?

stscieisenhamer committed:
Incorporate suggested copyedits and fix slideshow

stscieisenhamer commented:
I incorporated @eteq comments and fixed the slideshow bit.

eteq commented:
Awesome.  Looks good to me if @ivastar is also happy with it.

ivastar commented:
Looks good.

eteq commented:
Great - merging!
